Porting System.Windows.Forms.Design.ListControlStringCollectionEditor

vladimir-krestov opened the pull request:
Fixes #1271 
Related issue: #1115
Have you experienced this same bug with .NET Framework?: No

## Proposed changes

- Add System.Windows.Forms.Design.ListControlStringCollectionEditor class
- Add resources for DataSourceLocksItem
- Implement TypeForwardedTo method

<!-- We are in TELL-MODE the following section must be completed -->

## Customer Impact

- Changed collection editor for controls which use ListControlStringCollectionEditor (TextBox, ComboBox, ListBox, etc.) to compliance with .Net 4.8.

## Regression? 

- No

## Risk

- No

<!-- end TELL-MODE -->


## Screenshots <!-- Remove this section if PR does not change UI -->

### Before
- ComboBox (and others) has ObjectCollectionEditor as Items editor
![image](https://user-images.githubusercontent.com/49272759/66759214-81234700-eea8-11e9-9635-2ac442b7b29c.png)

<!-- TODO -->

### After
- ComboBox (and others) has ListControlStringCollectionEditor as Items editor 
![image](https://user-images.githubusercontent.com/49272759/66755564-e7a46700-eea0-11e9-8b2f-9d6f4eb1ae0c.png)

<!-- TODO -->


## Test methodology <!-- How did you ensure quality? -->

- Manual UI testing


 

## Test environment(s) <!-- Remove any that don't apply -->

- .Net Core version: 3.1.0-preview1.19458.7
- Microsoft Windows [Version 10.0.18362.418]


<!-- Mention language, UI scaling, or anything else that might be relevant -->


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2078)

vladimir-krestov commented:
Needs review before testing.

weltkante commented:
Its probably not enough to just add the editor, it must be linked up with the TypeDescriptor. We've had a whole bunch of other editor porting issues (e.g. #1119 and #1120) due to the editors not being linked up correctly. If possible we should do it correctly here in the first place.

vladimir-krestov commented:
@weltkante, thank you, I'll check it.

weltkante commented:
@vladimir-krestov Background: CoreFX System.Drawing doesn't want a direct reference to WinForms anymore so they removed some of the attributes required for this. You'll have to discuss whether the correct place is to add the missing in CoreFX (see quote below for the location) or if WinForms needs to have its own TypeDescriptor extension logic.

Copied from the comment in #1121:
>  CoreFx apparently has [some TypeDescriptor logic](https://github.com/dotnet/corefx/blob/master/src/System.ComponentModel.TypeConverter/src/System/ComponentModel/ReflectTypeDescriptionProvider.cs) added to .NET Core which extends e.g. the `Color` type with a `TypeConverter` instead of placing it directly on the type as an attribute, but you will need more logic to add missing Editor associations. If CoreFx doesn't want to do it then WinForms needs to have similar logic to extend the TypeDescriptor.
> 
> You'll probably want a bunch of tests which assert that `TypeDescriptor.GetEditor` returns the correct values for types in System.Drawing which are commonly used on Controls (Font, Color, Image, Icon, etc.)

weltkante commented:
@vladimir-krestov sorry, I was overly cautious, this will probably just work, `ListBox.Items` [already has](https://github.com/dotnet/winforms/blob/971343425f4e76dfc4b8fccfb87cbca278245fdc/src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs#L694) the right attribute.

vladimir-krestov commented:
@weltkante , anyway, thank you for your information.

zsd4yr commented on src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/ListControlStringCollectionEditor.cs:
```diff
@@ -0,0 +1,33 @@
+        public override object EditValue(ITypeDescriptorContext context, IServiceProvider provider, object value)
+        {
+            // If we're trying to edit the items in an object that has a DataSource set, throw an exception
+            //
```
nit please remove empty // line

vladimir-krestov replied:
Thanks, fixed.

zsd4yr commented on src/System.Windows.Forms.Design.Editors/src/System/Windows/Forms/Design/ListControlStringCollectionEditor.cs:
```diff
@@ -0,0 +1,33 @@
+            // If we're trying to edit the items in an object that has a DataSource set, throw an exception
+            //
+            ListControl control = context.Instance as ListControl;
+            if (control != null && control.DataSource != null)
```
nit: 
```suggestion
            if (control?.DataSource != null)
```

vladimir-krestov replied:
Thanks, fixed.

RussKie commented:
Just to clarify, because the editor was [originally defined in System.Design.dll](https://referencesource.microsoft.com/#System.Design/System/Windows/Forms/Design/ListControlStringCollectionEditor.cs) we need to set type forwarding.

We can resolve nits later.

RussKie commented:
📢  Please note due to a limited timeframe and the scope of work we are going to expediate all UI type editors ports into 3.1 and are going to ignore all stylistic review points.
All cleanup work is welcome on the master branch.

RussKie commented:
@vladimir-krestov please rebase to resolve the MC

vladimir-krestov commented:
Rebased to the latest release branch.
Fixed merge conflicts.

RussKie commented on src/System.Design/src/System.Design.Forwards.cs:
```diff
@@ -7,3 +7,4 @@
 [assembly: TypeForwardedTo(typeof(System.Windows.Forms.Design.ImageCollectionEditor))]
 [assembly: TypeForwardedTo(typeof(System.Windows.Forms.Design.StringArrayEditor))]
 [assembly: TypeForwardedTo(typeof(System.Windows.Forms.Design.StringCollectionEditor))]
+[assembly: TypeForwardedTo(typeof(System.Windows.Forms.Design.ListControlStringCollectionEditor))]
```
Please put it in alphabetic order

vladimir-krestov replied:
Okay

vladimir-krestov replied:
Thanks, fixed.

RussKie commented on src/System.Design/src/System.Design.Forwards.cs:
```diff
@@ -5,5 +5,6 @@
 using System.Runtime.CompilerServices;
 
 [assembly: TypeForwardedTo(typeof(System.Windows.Forms.Design.ImageCollectionEditor))]
+[assembly: TypeForwardedTo(typeof(System.Windows.Forms.Design.ListControlStringCollectionEditor))]
```
This change is not necessary, as I just found out - https://github.com/dotnet/winforms/pull/2089#issuecomment-542946415
Apologies for the confusion.

vladimir-krestov replied:
I disagree. I checked it using dev/3.1-uitypeeditors brunch. If control Items property has Editor attribute and we remove TypeForwardedTo for this Editor then this editor works incorrectly.
For example, DomainUpDownItemCollection has StringCollectionEditor as Editor attribute. If we remove StringCollectionEditor from TypeForwardedTo then DomainUpDown Items collection has ObjectCollection as editor. That is incorrect.
I think, in this case, we should have TypeForwardedTo for ListControlStringCollectionEditor. 
@zsd4yr could you please confirm it or explain it if I'm wrong?

zsd4yr replied:
TypeForwardedTo statements in the facades should be generated automatically by the [ResolveMatchingContract](https://github.com/dotnet/winforms/blob/70353c566c83a82879f294502319d9c8e1a35345/eng/FacadeAssemblies.props#L24) of GenFacades.

If it is not, then we have *lots and lots* of issues.

vladimir-krestov replied:
@zsd4yr, in your comment https://github.com/dotnet/winforms/pull/2089#issuecomment-542946415 about [System.Design.Forwards](https://github.com/dotnet/winforms/blob/release/3.1/src/System.Design/src/System.Design.Forwards.cs):
> I think you should remove that file as it does not do anything.

If we remove the file or comment StringCollectionEditor code line, we get ObjectCollectionEditor (for example in DomainUpDown).
![image](https://user-images.githubusercontent.com/49272759/67148959-36da0580-f2ae-11e9-8489-d257393c6842.png)
![image](https://user-images.githubusercontent.com/49272759/67148974-612bc300-f2ae-11e9-8a59-38641449d4a1.png)
- Before removing:
![image](https://user-images.githubusercontent.com/49272759/67149008-c67fb400-f2ae-11e9-8f8d-d4e394e6ad94.png)
- After:
![image](https://user-images.githubusercontent.com/49272759/67149146-70ac0b80-f2b0-11e9-93e9-24f95392db29.png)


Even if we do so, we still have ObjectCollectionEditor:
![image](https://user-images.githubusercontent.com/49272759/67149471-2e84c900-f2b4-11e9-8406-ca6fd9665565.png)

![image](https://user-images.githubusercontent.com/49272759/67149403-8a028700-f2b3-11e9-9972-494061a5a5dc.png)



So it seems we have _lots and lots_ of issues.
Please check it and make sure too.

RussKie replied:
Thank you @vladimir-krestov. I have spent today verifying the behaviour, and type forwarding does not appear to be necessary as pointed by @zsd4yr et al.

I have tested it on a brand new VM with 3.1.0-preview1.19506.1

| 3.1.0-preview1.19506.1 | Private binaries from bf30e621a68c93677bc1ceb694f5fa584b6eae95 |
|----|-----|
| ![image](https://user-images.githubusercontent.com/4403806/67181691-11253b80-f429-11e9-9228-e857fd4d90e0.png) | ![image](https://user-images.githubusercontent.com/4403806/67177899-542ce200-f41c-11e9-86b1-66e17cd1fb82.png) |
| ![image](https://user-images.githubusercontent.com/4403806/67181716-2c904680-f429-11e9-8e4e-32b2f2c6eac8.png) | ![image](https://user-images.githubusercontent.com/4403806/67177910-60b13a80-f41c-11e9-94de-fbdde890530f.png) |
| ![image](https://user-images.githubusercontent.com/4403806/67181826-6e20f180-f429-11e9-8607-ac25d567f7c6.png) | ![image](https://user-images.githubusercontent.com/4403806/67177921-6c046600-f41c-11e9-978f-715b7e3c8bda.png) |
| ![image](https://user-images.githubusercontent.com/4403806/67181945-b4765080-f429-11e9-8e32-071d47fd7b52.png) | ![image](https://user-images.githubusercontent.com/4403806/67178011-d7e6ce80-f41c-11e9-88c3-1cf4e9586c98.png) |
| ![image](https://user-images.githubusercontent.com/4403806/67182018-dcfe4a80-f429-11e9-9952-bc6239f4936d.png) | ![image](https://user-images.githubusercontent.com/4403806/67178064-209e8780-f41d-11e9-96f4-8198f452af01.png) | 

I have copied the following assemblies:
* Accessibility.dll
* System.Design.dll
* System.Drawing.Design.dll
* System.Drawing.dll
* System.Windows.Forms.Design.dll
* System.Windows.Forms.Design.Editors.dll
* System.Windows.Forms.dll

In your tests were you copying the same set of assemblies?
Without System.Design.dll / System.Drawing.Design.dll / System.Drawing.dll you won't see correct results.

zsd4yr replied:
^ this is correct. You also need the facades

RussKie replied:
Ok we tested further

1. With type forwarding - the way the current PR has it
2. Without type forwarding - System.Design.Forwards.cs deleted
3. Without type forwarding / manual registration - System.Design.Forwards.cs deleted, registered `System.Windows.Forms.Design.StringCollectionEditor` manually
```diff
diff --git a/src/System.Windows.Forms/src/System/Drawing/Design/UITypeEditor.cs b/src/System.Windows.Forms/src/System/Drawing/Design/UITypeEditor.cs
index 5c1471b8..ec7d20f2 100644
--- a/src/System.Windows.Forms/src/System/Drawing/Design/UITypeEditor.cs
+++ b/src/System.Windows.Forms/src/System/Drawing/Design/UITypeEditor.cs
@@ -6,6 +6,7 @@ using System.Collections;
 using System.Collections.ObjectModel;
 namespace System.Drawing.Design
 {
@@ -26,7 +27,9 @@ namespace System.Drawing.Design
-                [typeof(Collection<string>)] = "System.Windows.Forms.Design.StringCollectionEditor, " + AssemblyRef.SystemDesign
+                [typeof(Collection<string>)] = "System.Windows.Forms.Design.StringCollectionEditor, " + AssemblyRef.SystemDesign,
+                [typeof(DomainUpDown.DomainUpDownItemCollection)] = "System.Windows.Forms.Design.StringCollectionEditor, " + AssemblyRef.SystemDesign
             };
```


| Scenario 1 - ✔️ | Scenario 2 / 3 ❌  |
|---|----|
| ![image](https://user-images.githubusercontent.com/4403806/67262569-0b932880-f4f1-11e9-8363-0d6f119bdb90.png) | ![image](https://user-images.githubusercontent.com/4403806/67262779-abe94d00-f4f1-11e9-9940-58dd11f268f1.png) | 
| ![image](https://user-images.githubusercontent.com/4403806/67262599-2796ca00-f4f1-11e9-86aa-cca896e7053e.png) | ![image](https://user-images.githubusercontent.com/4403806/67262756-9b38d700-f4f1-11e9-84f2-4fd4dc612323.png) |

RussKie replied:
We've looked at the generated list of attributes in ILSpy and... neither `StringCollectionEditor`, nor `StringArrayEditor` nor `ListControlStringCollectionEditor` were there.... 😱 
![image](https://user-images.githubusercontent.com/4403806/67265171-d12d8980-f4f8-11e9-91de-1702f4838f3c.png)

@zsd4yr @dreddy-work we need your help, I've run out of hypothesis.

vladimir-krestov replied:
@zsd4yr, please check it independently in `release3.1-editors` branch, and if `StringCollectionEditor` works well for you please tell us steps to reproduce.

zsd4yr replied:
@RussKie @dreddy-work Per https://github.com/dotnet/winforms/pull/2092#issuecomment-545065158 Could this be happening because `StringCollectionEditor` is internal? Meaning that it is not present in the ref and so GenFacades does not know to generate a type-forward?

RussKie replied:
We tried making it `public` and inherit directly from `UITypeEditor` - neither helped.

zsd4yr replied:
No it appears to be present in the [targeting pack](https://dotnet.myget.org/feed/dotnet-core/package/nuget/Microsoft.TargetingPack.NETFramework.v4.7). But is it present in our refs? 

![image](https://user-images.githubusercontent.com/10677481/67336992-25297580-f4db-11e9-98f8-f5e3c600f623.png)

zsd4yr replied:
Interesting, it is **also** present in the ref lib for System.Windows.Forms.Editors. 

![image](https://user-images.githubusercontent.com/10677481/67337242-84878580-f4db-11e9-9cec-36f57a04dddf.png)

And ResolveMatchingContract is running for the facade and finding the correct project in the references.

![image](https://user-images.githubusercontent.com/10677481/67337637-4474d280-f4dc-11e9-95de-aebd28cf3915.png)

Somebody will have to dig into GenFacades! This is really where my knowledge ends...

zsd4yr replied:
@vladimir-krestov @RussKie just had a chat with @Shyam-Gupta and @dreddy-work about this. I am going to over-explain as it was helpful for me to be really verbose.

First, TypeForwardTo exists for public types only. For these types, we don't want to force customers to change their references, so instead we add these TypeForwardTo stubs which tell the CLI that in fact the types you are seeking are in another binary. We use these only for public types.

For internals, as is the case here and also with some other editors, the right thing to do is **not** to add TypeForwardTo. Instead the right thing to do is to change the location referencing this internal type to instead reference it in its new location. That would mean changing  any instance of `"System.Windows.Forms.Design.StringCollectionEditor, " + AssemblyRef.SystemDesign` in our code instead to `[typeof(Collection<string>)] = "System.Windows.Forms.Design.StringCollectionEditor, " + AssemblyRef.SystemWinformsDesign`. This is because customers should never access these internals and so it is not a concern of ours to ensure ease-of-access with a TypeForwardTo when types move assemblies. Instead, we should change our own implementation to access the new location of the internal.

weltkante replied:
> This is because customers should never access these internals and so it is not a concern of ours to ensure ease-of-access with a TypeForwardTo when types move assemblies.

I'm not objecting, but I'll mention that its possible to place `[Editor(...)]` attributes on user types referencing internal editors via string names. If you don't have type forwards those will have to be adjusted by users when they port their application (which IMHO is ok since they were referring to internals). Just wanted to mention it so a conscious decision can be made.

RussKie replied:
Thank you @weltkante, it is a good point.

zsd4yr replied:
This is a really good point. haha! Wow every time I talk to @ericstj I realize how little I truly know 😄 

Okay, so first off, philosophically my comment from 40 minutes ago is correct; however, for many reasons --one of which @weltkante pointed out-- this will not work for us. To be clear, we do not and we still cannot be expected to support customers who reference our internals. The public API is our contract. You are right however, that the Editor Attribute as we have used it and as others are able to use it has allowed these types, like `StringCollectionEditor`, to _leak_ into our public surface as strings which must be resolved in attributes over public types. 

That said, I think after talking to Eric, @Shyam-Gupta, and @dreddy-work --and taking the additional concerns in mind-- that the most conservative and safe approach is **to manually add these TypeForwardTo statements when we want to type forward internals**.

@RussKie my apologies then for telling you that we should not be doing this. We hope not to to this, and we should never have to manually add these for publics, but for internals which we have allowed to _leak_ into our public surface area via these editors, we need to manually maintain these TypeForwardTo statements in a file as you originally did.

RussKie replied:
Building and **maintaining** frameworks is hard 😆 

Great collaboration, thank you all.

RussKie replied:
@dreddy-work in light of the latest discoveries, do we need to revisit #1545 and setup type forwarding?

vladimir-krestov replied:
Thank you all for explanations.

zsd4yr replied:
> @dreddy-work in light of the latest discoveries, do we need to revisit #1545 and setup type forwarding?

I had the same concern. I have created #2171 and assigned it to @dreddy-work

zsd4yr replied:
I have also created #2172 to track the creation of a test for these as I suspect that we have mistakenly missed others.

merriemcgaw commented:
approved from my end. Please add the servicing: consider label after testing completes.

vladimir-krestov commented:
Rebased to `dev/3.1-uitypeeditors` branch.
Added `TypeForwardedTo` for `ListControlCollectionEditor` according to https://github.com/dotnet/winforms/pull/2078#discussion_r337787642

vladimir-krestov commented:
@RussKie and @zsd4yr I need your approval before testing.

RussKie commented on src/System.Design/src/System.Design.Forwards.cs:
```diff
@@ -0,0 +1,7 @@
+
+using System.Runtime.CompilerServices;
+
+[assembly: TypeForwardedTo(typeof(System.Windows.Forms.Design.ListControlStringCollectionEditor))]
```
I think we still need `StringCollectionEditor` and `ArrayEditor`...
